react/features/calendar-sync/web/microsoftCalendar.js

```diff
@@ -0,0 +1,564 @@
+                        'Popup closed before completing auth.');
+                    popupAuthWindow = null;
+                    window.removeEventListener('message', handleAuth);
+                    clearInterval(windowCloseCheck);
+                } else if (!popupAuthWindow) {
+                    // This case probably happened because the user completed
+                    // auth.
+                    clearInterval(windowCloseCheck);
+                }
+            }, 500);
+
+            /**
+             * Callback with scope access to other variables that are part of
+             * the sign in request.
+             *
+             * @param {Object} event - The event from the post message.
+             * @private
+             * @returns {void}
+             */
+            function handleAuth({ data }) {
+                if (!data || data.type !== 'ms-login') {
+                    return;
+                }
+
+                window.removeEventListener('message', handleAuth);
+
+                popupAuthWindow && popupAuthWindow.close();
+                popupAuthWindow = null;
+
+                const params = getParamsFromHash(data.hash);
+                const tokenParts = getValidatedTokenParts(
+                    params, guids, microsoftApiApplicationClientID);
+
+                if (!tokenParts) {
+                    signInDeferred.reject('Invalid token received');
+
+                    return;
+                }
+
+                dispatch(setCalendarAPIAuthState({
+                    authState: undefined,
+                    accessToken: tokenParts.accessToken,
+                    idToken: tokenParts.idToken,
+                    tokenExpires: tokenParts.tokenExpires,
+                    userDomainType: tokenParts.userDomainType,
+                    userSigninName: tokenParts.userSigninName
+                }));
+
+                signInDeferred.resolve();
+            }
+
+            window.addEventListener('message', handleAuth);
+
+            return signInDeferred.promise;
+        };
+    },
+
+    /**
+     * Sign out from the Microsoft API Client Library.
+     *
+     * @returns {function(Dispatch<*>): Promise<void>}
+     */
+    signOut(): Function {
+        // Using the Outlook API Get Started guide, there is no revoking to be
+        // done with the Microsoft API. Instead any stored state on the app-side
+        // should be cleared and the token should be allowed to expire.
+        return () => Promise.resolve();
+    },
+
+    /**
+     * Updates the profile data using calendar-sync feature.
+     *
+     * @returns {function(Dispatch<*>): Promise<void>}
+     */
+    updateProfile(): Function {
+        return (dispatch: Dispatch<*>, getState: Function) => {
+            const { msAuthState = {} }
+                = getState()['features/calendar-sync'] || {};
+            const email = msAuthState.userSigninName || '';
+
+            dispatch(setCalendarProfileEmail(email));
```

Comment:
This is kind of odd. I wonder if it's better to just make an extra call here to get the email or do more in the reducers. It just feels off to me that this action is essentially looking at state elsewhere in redux and duplicating it.